packages/data-api/src/TupaiaDataApi.js

```diff
@@ -129,30 +131,69 @@ export class TupaiaDataApi {
   async fetchDataElementsMetadataFromSqlQuery(sqlQuery, includeOptions) {
     const dataElementsMetadata = await sqlQuery.executeOnDatabase(this.database);
 
+    // includeOptions = true, should also fetch metadata for options from both question.options and question.option_set_id
     if (includeOptions) {
-      return dataElementsMetadata.map(({ options, type, ...restOfMetadata }) => {
-        return {
-          options: this.buildOptionsMetadata(options, type),
-          ...restOfMetadata, // type will not be included
-        };
-      });
+      // Get all possible option_set_ids from questions
+      const optionSetIds = [
+        ...new Set(dataElementsMetadata.filter(d => !!d.option_set_id).map(d => d.option_set_id)),
+      ];
+      // Get all the options from the option sets and grouped by set ids.
+      const optionsGroupedBySetId = await this.getOptionsGroupedBySetId(optionSetIds);
+
+      return dataElementsMetadata.map(
+        ({ options, type, option_set_id: optionSetId, ...restOfMetadata }) => {
+          // In reality, the options can only come from either question.options or question.option_set_id
+          const allOptions = [...options, ...(optionsGroupedBySetId[optionSetId] || [])];
+          return {
+            options: this.buildOptionsMetadata(allOptions, type),
+            ...restOfMetadata,
+          };
+        },
+      );
+    }
+
+    // includeOptions = false, return basic data elements metadata
+    return dataElementsMetadata.map(
+      ({ options, type, option_set_id: optionSetId, ...restOfMetadata }) => ({
+        ...restOfMetadata,
+      }),
+    );
+  }
+
+  async getOptionsGroupedBySetId(optionSetIds) {
+    if (!optionSetIds || !optionSetIds.length) {
+      return {};
     }
 
-    return dataElementsMetadata;
+    const options = await new SqlQuery(
+      `
+      SELECT option.value, option.label, option.option_set_id
+      FROM option
+      WHERE option.option_set_id IN ${SqlQuery.parameteriseArray(optionSetIds)}
+      `,
+      optionSetIds,
+    ).executeOnDatabase(this.database);
+
+    return groupBy(options, 'option_set_id');
   }
 
   buildOptionsMetadata = (options = [], type) => {
     const optionList = options.length === 0 && type === 'Binary' ? DEFAULT_BINARY_OPTIONS : options;
     const optionsMetadata = {};
+
     optionList.forEach(option => {
       try {
-        const { value, label } = JSON.parse(option);
+        // options coming from question.options are JSON format strings
+        // options coming from option_set are actual JSON objects
+        const optionObject = typeof option === 'string' ? JSON.parse(option) : option;
```

Comment:
This function doesn't otherwise know about the source of options, so I'd prefer to keep it ignorant. Could you instead (in the succinct line earlier) `map` and `parse` the options at the point they're gathered from each source?